browser_use/agent/service.py

```diff
@@ -396,27 +394,78 @@ def _setup_action_models(self) -> None:
 		self.DoneActionModel = self.controller.registry.create_action_model(include_actions=['done'])
 		self.DoneAgentOutput = AgentOutput.type_with_custom_actions(self.DoneActionModel)
 
-	def _set_tool_calling_method(self) -> ToolCallingMethod | None:
-		tool_calling_method = self.settings.tool_calling_method
-		if tool_calling_method == 'auto':
-			if is_model_without_tool_support(self.model_name):
-				return 'raw'
-			elif self.chat_model_library == 'ChatGoogleGenerativeAI':
-				return None
-			elif self.chat_model_library == 'ChatOpenAI':
-				return 'function_calling'
-			elif self.chat_model_library == 'AzureChatOpenAI':
-				# Azure OpenAI API requires 'tools' parameter for GPT-4
-				# The error 'content must be either a string or an array' occurs when
-				# the API expects a tools array but gets something else
-				if 'gpt-4' in self.model_name.lower():
-					return 'tools'
-				else:
-					return 'function_calling'
+	def _test_tool_calling_method(self, method: str) -> bool:
+		"""Test if a specific tool calling method works with the current LLM."""
+		try:
+			# Create a simple test message
+			test_message = HumanMessage(content='Hello, world!')
+
+			if method == 'raw':
+				# For raw mode, we just check if we can invoke the model
```

Comment:
the test for `raw` should probably ask the model to respond with some JSON manually in the same way we do in the actual system prompt in `raw` mode.

I think for the same reason we check if the model knows the capital of France, basic sanity checks are very useful to weed out low power models that struggle to follow directions + return JSON